File: javascript/packages/linter/src/rules/erb-no-trailing-whitespace.ts
import { Location, Visitor } from "@herb-tools/core"
import { ParserRule, Mutable, BaseAutofixContext } from "../types.js"

import { isHTMLOpenTagNode, isHTMLTextNode, isLiteralNode, isWhitespaceNode, getTagLocalName } from "@herb-tools/core"
import { findNodeAtPosition } from "./rule-utils.js"

import type { UnboundLintOffense, LintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLElementNode, HTMLTextNode, LiteralNode, WhitespaceNode, ParseResult, DocumentNode, ERBNode } from "@herb-tools/core"

const WHITESPACE_CHARACTERS = new Set([" ", "\t", "\r", "\v", "\f", "\u00A0"])

function trailingWhitespaceStart(value: string): number {
  let start = value.length

  while (start > 0 && WHITESPACE_CHARACTERS.has(value[start - 1])) start--

  return start
}

function trimTrailingWhitespace(value: string): string {
  const start = trailingWhitespaceStart(value)

  return start === value.length ? value : value.slice(0, start)
}

function trimWhitespaceBeforeNewlines(value: string): string {
  if (!value.includes("\n")) return value

  const segments = value.split("\n")

  for (let index = 0; index < segments.length - 1; index++) {
    segments[index] = trimTrailingWhitespace(segments[index])
  }

  return segments.join("\n")
}

function isOnlyWhitespace(value: string): boolean {
  return value.length > 0 && trailingWhitespaceStart(value) === 0
}

interface SkipZone {
  startLine: number
  startColumn: number
  endLine: number
  endColumn: number
}

interface TrailingWhitespaceCandidate {
  line: number
  column: number
  length: number
}

interface ERBNoTrailingWhitespaceAutofixContext extends BaseAutofixContext {
  node: Mutable<HTMLTextNode> | Mutable<LiteralNode> | Mutable<WhitespaceNode>
}

class SkipZoneCollector extends Visitor {
  skipZones: SkipZone[] = []

  SKIP_TAGS = new Set(["pre", "textarea", "script", "style"])

  visitHTMLElementNode(node: HTMLElementNode): void {
    if (isHTMLOpenTagNode(node.open_tag)) {
      const tagName = getTagLocalName(node.open_tag)

      if (tagName && this.SKIP_TAGS.has(tagName)) {
        this.skipZones.push({
          startLine: node.location.start.line,
          startColumn: node.location.start.column,
          endLine: node.location.end.line,
          endColumn: node.location.end.column
        })

        return
      }
    }

    super.visitHTMLElementNode(node)
  }

  visitERBNode(node: ERBNode) {
    if (!node.tag_opening) return
    if (!node.tag_closing) return

    this.skipZones.push({
      startLine: node.tag_opening.location.start.line,
      startColumn: node.tag_opening.location.start.column,
      endLine: node.tag_closing.location.end.line,
      endColumn: node.tag_closing.location.end.column
    })
  }
}

export class ERBNoTrailingWhitespaceRule extends ParserRule<ERBNoTrailingWhitespaceAutofixContext> {
  static autocorrectable = true
  static ruleName = "erb-no-trailing-whitespace"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: {
        cli: "error",
        editor: "info",
      },
    }
  }

  check(result: ParseResult, _context?: Partial<LintContext>): UnboundLintOffense<ERBNoTrailingWhitespaceAutofixContext>[] {
    const offenses: UnboundLintOffense<ERBNoTrailingWhitespaceAutofixContext>[] = []
    const lines = result.source.split("\n")
    const candidates = this.findTrailingWhitespaceCandidates(lines)

    if (candidates.length === 0) return offenses

    const skipZones = this.collectSkipZones(result.value)

    for (const candidate of candidates) {
      if (!this.isInSkipZone(candidate, skipZones)) {
        const location = Location.from(candidate.line, candidate.column, candidate.line, candidate.column + candidate.length)
        const node = findNodeAtPosition(result.value, candidate.line, candidate.column, (n) => isHTMLTextNode(n) || isLiteralNode(n) || isWhitespaceNode(n)) as HTMLTextNode | LiteralNode | WhitespaceNode | null

        offenses.push(this.createOffense("Extra whitespace detected at end of line.", location, node ? { node } : undefined))
      }
    }

    return offenses
  }

  private findTrailingWhitespaceCandidates(lines: string[]): TrailingWhitespaceCandidate[] {
    const candidates: TrailingWhitespaceCandidate[] = []

    for (let i = 0; i < lines.length; i++) {
      const line = lines[i]
      const start = trailingWhitespaceStart(line)

      if (start < line.length) {
        candidates.push({
          line: i + 1,
          column: [...line.slice(0, start)].length,
          length: [...line.slice(start)].length
        })
      }
    }

    return candidates
  }

  private collectSkipZones(root: DocumentNode): SkipZone[] {
    const collector = new SkipZoneCollector()

    collector.visit(root)

    return collector.skipZones
  }

  private isInSkipZone(candidate: TrailingWhitespaceCandidate, skipZones: SkipZone[]): boolean {
    for (const zone of skipZones) {
      if (candidate.line < zone.startLine || candidate.line > zone.endLine) continue
      if (candidate.line === zone.endLine && candidate.column >= zone.endColumn) continue
      if (candidate.line === zone.startLine && candidate.column < zone.startColumn) continue

      return true
    }

    return false
  }

  autofix(offense: LintOffense<ERBNoTrailingWhitespaceAutofixContext>, result: ParseResult, _context?: Partial<LintContext>): ParseResult | null {
    if (!offense.autofixContext) return null

    const { node } = offense.autofixContext

    if (node.type === "AST_WHITESPACE_NODE") {
      return this.autofixWhitespaceNode(offense, node, result)
    }

    if (node.type === "AST_HTML_TEXT_NODE" || node.type === "AST_LITERAL_NODE") {
      let fixedContent = trimWhitespaceBeforeNewlines(node.content)
      const offenseIsAtEndOfContent = this.isOffenseAtEndOfContent(offense, node)

      if (offenseIsAtEndOfContent) {
        if (this.hasTrailingWhitespaceNotIndentation(fixedContent)) {
          fixedContent = trimTrailingWhitespace(fixedContent)
        }

        if (isOnlyWhitespace(fixedContent) && node.location.start.column !== 0) {
          fixedContent = ""
        }
      }

      node.content = fixedContent
    }

    return result
  }

  private autofixWhitespaceNode(offense: LintOffense<ERBNoTrailingWhitespaceAutofixContext>, node: Mutable<WhitespaceNode>, result: ParseResult): ParseResult | null {
    if (!node.value) return null

    const originalValue = node.value.value
    let fixedValue = trimWhitespaceBeforeNewlines(originalValue)

    if (this.isOffenseAtEndOfContent(offense, node)) {
      fixedValue = trimTrailingWhitespace(fixedValue)
    }

    if (fixedValue === originalValue) return null

    node.value.value = fixedValue

    return result
  }

  private isOffenseAtEndOfContent(offense: LintOffense<ERBNoTrailingWhitespaceAutofixContext>, node: Mutable<HTMLTextNode> | Mutable<LiteralNode> | Mutable<WhitespaceNode>): boolean {
    return offense.location.end.line === node.location.end.line && offense.location.end.column === node.location.end.column
  }

  private hasTrailingWhitespaceNotIndentation(content: string): boolean {
    if (content.endsWith("\n")) return false

    const whitespaceStart = trailingWhitespaceStart(content)
    if (whitespaceStart === content.length) return false
    if (whitespaceStart === 0) return false

    const characterBefore = content[whitespaceStart - 1]
    if (characterBefore === "\n") return false

    return true
  }
}

File: rust/herb-linter/src/rules/erb_no_trailing_whitespace.rs
use crate::offense::Offense;
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use herb::nodes::{AnyNode, DocumentNode};

use herb::nodes::*;
use herb::union_types::*;
use herb::Location;
use herb::ParseResult;
use herb::Visitor;

use herb_config::{Severity, SeverityConfig};

pub struct ERBNoTrailingWhitespaceRule;

struct SkipZone {
  start_line: u32,
  start_column: u32,
  end_line: u32,
  end_column: u32,
}

struct TrailingWhitespaceCandidate {
  line: u32,
  column: u32,
  length: u32,
}

struct SkipZoneCollector {
  skip_zones: Vec<SkipZone>,
}

impl SkipZoneCollector {
  fn is_skip_tag(tag_name: &str) -> bool {
    matches!(tag_name.to_lowercase().as_str(), "pre" | "textarea" | "script" | "style")
  }
}

impl Visitor for SkipZoneCollector {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if let Some(ref open_tag) = node.open_tag {
      match open_tag {
        ERBOpenTagNodeOrHTMLConditionalOpenTagNodeOrHTMLOpenTagNode::HTMLOpenTagNode(tag) => {
          if let Some(ref tag_name) = tag.tag_name {
            if Self::is_skip_tag(&tag_name.value) {
              self.skip_zones.push(SkipZone {
                start_line: node.location.start.line,
                start_column: node.location.start.column,
                end_line: node.location.end.line,
                end_column: node.location.end.column,
              });

              return;
            }
          }
        }
        _ => {}
      }
    }

    self.walk_html_element_node(node);
  }

  fn visit_erb_node(&mut self, node: &dyn ERBNode) {
    if let (Some(tag_opening), Some(tag_closing)) = (node.tag_opening(), node.tag_closing()) {
      self.skip_zones.push(SkipZone {
        start_line: tag_opening.location.start.line,
        start_column: tag_opening.location.start.column,
        end_line: tag_closing.location.end.line,
        end_column: tag_closing.location.end.column,
      });
    }
  }
}

fn find_trailing_whitespace_candidates(lines: &[&str]) -> Vec<TrailingWhitespaceCandidate> {
  let mut candidates = Vec::new();

  for (index, line) in lines.iter().enumerate() {
    let trimmed_length = line
      .trim_end_matches(|character: char| {
        character == ' ' || character == '\t' || character == '\r' || character == '\x0B' || character == '\x0C' || character == '\u{00A0}'
      })
      .len();

    if trimmed_length < line.len() {
      candidates.push(TrailingWhitespaceCandidate {
        line: (index + 1) as u32,
        column: trimmed_length as u32,
        length: (line.len() - trimmed_length) as u32,
      });
    }
  }

  candidates
}

fn is_in_skip_zone(candidate: &TrailingWhitespaceCandidate, skip_zones: &[SkipZone]) -> bool {
  for zone in skip_zones {
    if candidate.line < zone.start_line || candidate.line > zone.end_line {
      continue;
    }

    if candidate.line == zone.end_line && candidate.column >= zone.end_column {
      continue;
    }

    if candidate.line == zone.start_line && candidate.column < zone.start_column {
      continue;
    }

    return true;
  }

  false
}

impl Rule for ERBNoTrailingWhitespaceRule {
  fn has_autofix(&self) -> bool {
    true
  }

  fn autocorrectable(&self) -> bool {
    true
  }

  fn name(&self) -> &'static str {
    "erb-no-trailing-whitespace"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("0.9.0")
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::Severity(Severity::Error)
  }
}

impl ParserRule for ERBNoTrailingWhitespaceRule {
  fn autofix(&self, offense: &Offense, document: &mut DocumentNode, context: &LintContext) -> bool {
    autofix(offense, document, context)
  }

  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    let lines: Vec<&str> = result.source.split('\n').collect();
    let candidates = find_trailing_whitespace_candidates(&lines);

    if candidates.is_empty() {
      return Vec::new();
    }

    let mut offenses = Vec::new();
    let mut collector = SkipZoneCollector { skip_zones: Vec::new() };

    collector.visit_document_node(&result.value);

    for candidate in &candidates {
      if !is_in_skip_zone(candidate, &collector.skip_zones) {
        let location = Location::from(candidate.line, candidate.column, candidate.line, candidate.column + candidate.length);

        offenses.push(UnboundOffense::new(self.name(), "Extra whitespace detected at end of line.", location));
      }
    }

    offenses
  }
}

const TRAILING: &[char] = &[' ', '\t', '\r', '\u{b}', '\u{c}', '\u{a0}'];

/// Drops horizontal whitespace that sits immediately before a newline.
fn strip_before_newlines(content: &str) -> String {
  content
    .split('\n')
    .collect::<Vec<_>>()
    .iter()
    .enumerate()
    .map(|(index, line)| {
      if index + 1 == content.split('\n').count() {
        line.to_string()
      } else {
        line.trim_end_matches(TRAILING).to_string()
      }
    })
    .collect::<Vec<_>>()
    .join("\n")
}

fn has_trailing_whitespace_not_indentation(content: &str) -> bool {
  if content.ends_with('\n') {
    return false;
  }

  let trimmed = content.trim_end_matches(TRAILING);

  if trimmed.len() == content.len() || trimmed.is_empty() {
    return false;
  }

  !trimmed.ends_with('\n')
}

fn autofix(offense: &Offense, document: &mut DocumentNode, _context: &LintContext) -> bool {
  let mut fixed = false;

  crate::autofix::walk_nodes_mut(&mut document.children, &mut |node| {
    let (content, location) = match node {
      AnyNode::HTMLTextNode(text) => (&mut text.content, &text.location),
      AnyNode::LiteralNode(literal) => (&mut literal.content, &literal.location),
      _ => return,
    };

    let within = offense.location.start.line >= location.start.line && offense.location.end.line <= location.end.line;

    if !within {
      return;
    }

    let at_end_of_content = offense.location.end.line == location.end.line && offense.location.end.column == location.end.column;

    let starts_at_column_zero = location.start.column == 0;
    let mut updated = strip_before_newlines(content);

    if at_end_of_content {
      if has_trailing_whitespace_not_indentation(&updated) {
        updated = updated.trim_end_matches(TRAILING).to_string();
      }

      if !updated.is_empty() && updated.chars().all(|character| TRAILING.contains(&character)) && !starts_at_column_zero {
        updated = String::new();
      }
    }

    *content = updated;

    // the JavaScript rule reports the offense as fixed whenever it owns the node,
    // even when an earlier offense already cleaned the content
    fixed = true;
  });

  fixed
}
